src/MLabsPAB/PreBalance.hs

```diff
@@ -54,41 +54,65 @@ preBalanceTxIO ::
   PubKeyHash ->
   UnbalancedTx ->
   Eff effs (Either Text Tx)
-preBalanceTxIO pabConf ownPkh tx =
+preBalanceTxIO pabConf ownPkh unbalancedTx =
   runEitherT $
     do
       utxos <- lift $ CardanoCLI.utxosAt pabConf $ Ledger.pubKeyHashAddress ownPkh
       privKeys <- newEitherT $ Files.readPrivateKeys pabConf
-      let utxoIndex = fmap Tx.toTxOut utxos <> unBalancedTxUtxoIndex tx
-          tx' = unBalancedTxTx tx
-          requiredSigs = Map.keys (unBalancedTxRequiredSignatories tx)
+      let utxoIndex = fmap Tx.toTxOut utxos <> unBalancedTxUtxoIndex unbalancedTx
+          tx = unBalancedTxTx unbalancedTx
+          requiredSigs = Map.keys (unBalancedTxRequiredSignatories unbalancedTx)
 
-      minUtxo <- newEitherT $ CardanoCLI.calculateMinUtxo pabConf tx
+      lift $ printLog Debug $ show utxoIndex
+
+      loop utxoIndex privKeys requiredSigs [] tx
+  where
+    loop utxoIndex privKeys requiredSigs prevMinUtxos tx = do
+      nextMinUtxos <-
+        newEitherT $
+          calculateMinUtxos pabConf $ filter (`notElem` map fst prevMinUtxos) $ Tx.txOutputs tx
```

Comment:
Could filter by notElem be replaced with something like list difference [\\\\](https://hackage.haskell.org/package/base-4.16.0.0/docs/Data-List.html#v:-92--92-)